From eb2ee5b4856adf3360e1397c45c457594f7b6ead Mon Sep 17 00:00:00 2001
From: Icenowy Zheng <uwu@icenowy.me>
Date: Sat, 6 Jul 2024 00:10:39 +0800
Subject: [PATCH 21/25] gallivm: add LoongArch support to the mattrs setting
 code

Currently the mattrs is set according to the softdev convention, with
LSX explicitly disabled because it's troublesome at least on LLVM 17.

Signed-off-by: Icenowy Zheng <uwu@icenowy.me>
---
 src/gallium/auxiliary/gallivm/lp_bld_misc.cpp | 17 +++++++++++++++++
 1 file changed, 17 insertions(+)

diff --git a/src/gallium/auxiliary/gallivm/lp_bld_misc.cpp b/src/gallium/auxiliary/gallivm/lp_bld_misc.cpp
index f0e55bc4d46..07f6b8326ed 100644
--- a/src/gallium/auxiliary/gallivm/lp_bld_misc.cpp
+++ b/src/gallium/auxiliary/gallivm/lp_bld_misc.cpp
@@ -414,6 +414,23 @@ lp_build_fill_mattrs(std::vector<std::string> &MAttrs)
     */
    MAttrs = {"+m","+c","+a","+d","+f"};
 #endif
+
+#if DETECT_ARCH_LOONGARCH64 == 1
+   /*
+    * TODO: Implement util_get_cpu_caps()
+    *
+    * No FPU-less LoongArch64 systems are ever shipped yet, and LP64D is
+    * the default ABI, so FPU is enabled here.
+    *
+    * The Software development convention defaults to have "128-bit
+    * vector", so LSX is enabled here, see
+    * https://github.com/loongson/la-softdev-convention/releases/download/v0.1/la-softdev-convention.pdf
+    *
+    * FIXME: lsx is disabled here now because it's broken on LLVM17, need to
+    * test on LLVM18 now.
+    */
+   MAttrs = {"+f","+d","-lsx"};
+#endif
 }
 
 void
-- 
2.45.2

